Add support for vTPM and its runtime protocol

cclaudio opened the pull request:
This is a sketch of how the vTPM and the vTPM protocol would work in the SVSM, for now running in CPL0. 

The vTPM is initialized (manufactured) on every boot. If OVMF and the Linux guest kernel have proper tpm drivers, they should be able to communicate with the SVSM vTPM. See the section Dependencies below.

Once this is merged I can post a PR for the attestation protocol.

## Dependencies

OVMF
```
$ git clone https://github.com/coconut-svsm/edk2.git
$ cd edk2/
$ git fetch origin pull/2/head:edk2-pr2
$ git checkout edk2-pr2
$ git submodule init
$ git submodule update

$ export PYTHON3_ENABLE=TRUE
$ export PYTHON_COMMAND=python3
$ make -j16 -C BaseTools/
$ source ./edksetup.sh --reconfig
$ build -a X64 -b DEBUG -t GCC5 -D DEBUG_ON_SERIAL_PORT -D DEBUG_VERBOSE -DTPM2_ENABLE -p OvmfPkg/OvmfPkgX64.dsc
```

Guest kernel:

The linux branch below is based on the [PR#2](https://github.com/coconut-svsm/linux/pull/2) posted against the coconut-svsm/linux repository with additional two patches for the TPM platform driver.

`CONFIG_TCG_PLATFORM` should be enabled in the guest. The same kernel also needs to be installed on the host.

```
$ git clone https://github.com/cclaudio/linux.git
$ cd linux
$ git checkout svsm-host-guest-vtpm
```

QEMU:

Follow the instructions described in [INSTALL.md](https://github.com/coconut-svsm/svsm/blob/main/Documentation/INSTALL.md#building-qemu) to build the QEMU branch below.

```
$ git clone https://github.com/coconut-svsm/qemu.git
$ cd qemu
$ git fetch origin pull/2/head:qemu-pr2
$ git co qemu-pr2
```

## How to test it

Follow the [INSTALL.md](https://github.com/coconut-svsm/svsm/blob/main/Documentation/INSTALL.md#building-the-coconut-svsm) to build this PR and run a QEMU guest.

During the boot you should be able to see messages from OVMF and the linux kernel like these:

```
Loading PEIM at 0x0007F7FC000 EntryPoint=0x0007F7FD906 Tcg2ConfigPei.efi
Found SVSM TPM
Tcg2ConfigPeimEntryPoint
Tcg2ConfigPeimEntryPoint: TPM2 detected

[...]

ReadPcr - HashAlg = 0x0004, Pcr[00], digest = 51 C3 23 DE 0C 0C 69 4F 46 01 CD D0 2B EB 58 FF 13 62 9F 74 
ReadPcr - HashAlg = 0x000B, Pcr[00], digest = FC EC B5 6A CC 30 38 62 B3 0E B3 42 C4 99 0B EB 50 B5 E0 AB 89 72 24 49 C2 D9 A7 3F 37 B0 19 FE 
ReadPcr - HashAlg = 0x000C, Pcr[00], digest = 61 93 87 2D C7 23 D5 33 E3 BB 45 FB 0A EE C1 35 48 AD DE 71 11 DF 93 A4 D7 0C B1 B5 77 CE 31 10 4A C9 DF BC B8 76 BD 07 F7 7D 2C E4 B3 F7 33 DF

[...]

[    0.907842] SEV: SNP guest platform device initialized.
[    0.910991] SEV: SNP SVSM VTPM platform device initialized
[    0.993792] tpm tpm: TPM 2.0 platform device

```

From guest userspace you can also install the `tpm2-tools` package and use it to communicate with the SVSM-vTPM (e.g. `tpm2_pcrread` without arguments)

cclaudio commented:
Sorry for the branch conflict. I will fix it as soon as I find a new way to get the slab item size in the function `mm::alloc::layout_from_ptr()`.  Recently, 8fd4c80ec42939bdad49c3a8c5509cf60cce8da2 removed the Slab pointer from SlabPageInfo.

00xc commented:
> Sorry for the branch conflict. I will fix it as soon as I find a new way to get the slab item size in the function `mm::alloc::layout_from_ptr()`. Recently, [8fd4c80](https://github.com/coconut-svsm/svsm/commit/8fd4c80ec42939bdad49c3a8c5509cf60cce8da2) removed the Slab pointer from SlabPageInfo.

Feel free to revert that commit by including the slab size instead of the pointer.

Freax13 commented on src/vtpm/wrapper.rs:
```diff
@@ -0,0 +1,66 @@
+    ptr::null_mut()
+}
+
+#[no_mangle]
+pub extern "C" fn realloc(p: *mut c_void, size: c_ulong) -> *mut c_void {
```
This function, `free` and `serial_out` should be marked as unsafe.

cclaudio replied:
Done, thanks!

cclaudio commented:
- Added code to cache the TPM EK in the TSS format (marshaled TPMT_PUBLIC). It needs to be provided when an attestation service is requested at runtime.
- Reverted https://github.com/coconut-svsm/svsm/commit/8fd4c80ec42939bdad49c3a8c5509cf60cce8da2
- Rebased to the latest main branch

cclaudio commented:
This PR version works with the upstream Microsoft TPM and OpenSSL. So, for testing, you only need to the OVMF and guest kernel mentioned in the PR description.

stefano-garzarella commented:
@cclaudio thanks for this great work! I'll review in the next days, in the meantime I report some building issues I'm having.

> The INSTALL.md is updated in this PR to reflect the steps above.

I can't see these changes (e.g. git submodule initialization), maybe we forgot them during rebase.

I think we should also update our container image with something like this:
```patch
diff --git a/scripts/container/opensuse-rust.docker b/scripts/container/opensuse-rust.docker
index bf8a109..ad276d6 100644
--- a/scripts/container/opensuse-rust.docker
+++ b/scripts/container/opensuse-rust.docker
@@ -19,7 +19,9 @@ ARG USER_NAME user
 SHELL ["/bin/bash", "-c"]
 
 RUN zypper ref && \
-    zypper install -y system-user-mail make gcc curl && \
+    zypper install -y system-user-mail make gcc curl \
+    patterns-devel-base-devel_basis \
+    autoconf autoconf-archive pkg-config automake libopenssl-devel && \
     useradd -u $USER_ID -m $USER_NAME
 
 USER $USER_NAME
```

This doesn't work for me, though, because I keep seeing:
```
(cd /home/stefano/repos/coconut/svsm-review/libvtpm/ms-tpm-20-ref/TPMCmd && \
	sed -i '/^AX_PTHREAD/d' configure.ac && \
	./bootstrap && \
	./configure \
		CFLAGS="-static -nostdinc -fno-stack-protector -fPIE -DSIMULATION=NO -DFILE_BACKED_NV=NO -I/home/stefano/repos/coconut/svsm-review/libvtpm/libcrt/include -I/home/stefano/repos/coconut/svsm-review/libvtpm/openssl/include" \
		LIBCRYPTO_LIBS="")
Generating file lists: src.mk
Setting up build
Setting default vendor strings
checking for gcc... gcc
checking whether the C compiler works... no
```

I think I need some help from OpenSUSE users :-)

stefano-garzarella commented on libvtpm/Makefile:
```diff
@@ -0,0 +1,140 @@
+CLANG_FLAGS = -Wno-incompatible-library-redeclaration
+
+$(BINDGEN):
+	cargo install bindgen-cli
```
On my system, I had an older version installed `bindgen v0.60.1` and after some struggle, I realized that this command was not updating it and I had several errors, so I suggest setting a version here:

```suggestion
	cargo install bindgen-cli@~0.69
```

cclaudio replied:
Makes sense. Done, thanks.

stefano-garzarella commented:
@roy-hopkins yeah thanks!
Adding `glibc-devel-static` and `git` worked (partially, I have other issues related to git-workspaces and bindgen), so the final changes are something like this:

```diff
diff --git a/scripts/container/opensuse-rust.docker b/scripts/container/opensuse-rust.docker
index bf8a109..1c05edd 100644
--- a/scripts/container/opensuse-rust.docker
+++ b/scripts/container/opensuse-rust.docker
@@ -19,7 +19,9 @@ ARG USER_NAME user
 SHELL ["/bin/bash", "-c"]
 
 RUN zypper ref && \
-    zypper install -y system-user-mail make gcc curl && \
+    zypper install -y system-user-mail make gcc curl \
+    patterns-devel-base-devel_basis glibc-devel-static git \
+    autoconf autoconf-archive pkg-config automake libopenssl-devel && \
     useradd -u $USER_ID -m $USER_NAME
 
 USER $USER_NAME
```

stefano-garzarella commented on libvtpm/Makefile:
```diff
@@ -0,0 +1,140 @@
+	$(MAKE) -C $(OPENSSL_DIR) -j$$(nproc)
+
+$(OPENSSL_MAKEFILE):
+	(cd $(OPENSSL_DIR) && git checkout OpenSSL_1_1_1q  && \
```
Using `./scripts/container/build.sh` in a `svsm` working tree created with `git-worktree`, I have the following issue:
```
(cd /home/stefano/repos/coconut/svsm-review/libvtpm/openssl && git checkout OpenSSL_1_1_1q  && \
	./Configure \
		--config=/home/stefano/repos/coconut/svsm-review/libvtpm/openssl_svsm.conf \
		SVSM \
		no-afalgeng \
		no-async \
		no-autoerrinit \
		no-autoload-config \
		no-bf \
		no-blake2 \
		no-capieng \
		no-cast \
		no-chacha \
		no-cms \
		no-ct \
		no-deprecated \
		no-des \
		no-dgram \
		no-dsa \
		no-dynamic-engine \
		no-ec2m \
		no-engine \
		no-err \
		no-filenames \
		no-gost \
		no-hw \
		no-idea \
		no-md4 \
		no-mdc2 \
		no-pic \
		no-ocb \
		no-poly1305 \
		no-posix-io \
		no-rc2 \
		no-rc4 \
		no-rfc3779 \
		no-rmd160 \
		no-scrypt \
		no-seed \
		no-sock \
		no-srp \
		no-ssl \
		no-stdio \
		no-threads \
		no-ts \
		no-whirlpool \
		no-shared \
		no-sse2 \
		no-ui-console \
		no-asm \
		--with-rand-seed=getrandom \
		-I/home/stefano/repos/coconut/svsm-review/libvtpm/libcrt/include \
		-Wl,rpath=/home/stefano/repos/coconut/svsm-review/libvtpm/libcrt -lcrt -g -O0)
fatal: not a git repository: /home/stefano/repos/coconut/svsm-review/libvtpm/openssl/../../../svsm/.git/worktrees/svsm-review1/modules/libvtpm/openssl
make[1]: *** [Makefile:35: /home/stefano/repos/coconut/svsm-review/libvtpm/openssl/Makefile] Error 128
make[1]: Leaving directory '/home/stefano/repos/coconut/svsm-review/libvtpm'
make: *** [Makefile:70: libvtpm/bindings.rs] Error 2
```

I suggest removing the `git checkout` here, since the git-submodules should already be at the right commit we have in the history:
```suggestion
	(cd $(OPENSSL_DIR) && \
```

This fixes the issue I saw, but I don't know if there is a better way.

cclaudio replied:
Actually, the `Subproject commit` I have for the openssl submodule is already the OpenSSL_1_1_1q. You are right, I can safely just remove the `git checkout` from the Makefile.

cclaudio commented on libvtpm/Makefile:
```diff
@@ -0,0 +1,140 @@
+$(LIBPLATFORM): $(MSTPM_MAKEFILE) $(LIBCRYPTO)
+	$(MAKE) -C $(MSTPM_DIR) $(LIBPLATFORM_A)
+
+MSTPM_CFLAGS := -static -nostdinc -fno-stack-protector -fPIE
+MSTPM_CFLAGS += -DSIMULATION=NO -DFILE_BACKED_NV=NO
+MSTPM_CFLAGS += -I$(LIBCRT_DIR)/include
+MSTPM_CFLAGS += -I$(OPENSSL_DIR)/include
```
Hmm, I think we should also pass `-mno-sse` to build openssl in `libvtpm/openssl_svsm.conf`. Let me check that

roy-hopkins replied:
The call to `Configure` in this `Makefile` already specifies `no-sse2` for openssl. I'm not sure if that propagates through to the compiler flags though.

cclaudio replied:
Openssl fails to build with `-mno-sse`, so I added the parameter only for the MS TPM build. Thanks
```
crypto/bio/b_print.c: In function 'abs_val':
crypto/bio/b_print.c:516:1: error: SSE register return with SSE disabled
 {
 ^
make[3]: *** [Makefile:1314: crypto/bio/b_print.o] Error 1
```

cclaudio commented:
Feedback applied

stefano-garzarella commented on Makefile:
```diff
@@ -53,14 +58,23 @@ bin/coconut-hyperv.igvm: $(IGVMBUILDER) stage1/kernel.elf stage1/stage2.bin
+src/vtpm/mstpm/bindings.rs: libvtpm/bindings.rs
+	cp libvtpm/bindings.rs $@
+
+libvtpm/bindings.rs:
```
I think this target conflicts with `libvtpm/libvtpm.a:`, because running with `make -j10` I have several errors:
```
checking for suffix of object files... make[1]: *** [Makefile:105: /home/stefano/repos/coconut/svsm-review/libvtpm/ms-tpm-20-ref/TPMCmd/Makefile] Error 77
make[1]: Leaving directory '/home/stefano/repos/coconut/svsm-review/libvtpm'
make: *** [Makefile:76: libvtpm/bindings.rs] Error 2
make: *** Waiting for unfinished jobs....
sed: can't read conftest.c: No such file or directory
configure: error: in `/home/stefano/repos/coconut/svsm-review/libvtpm/ms-tpm-20-ref/TPMCmd':
configure: error: cannot compute suffix of object files: cannot compile
See `config.log' for more details
make[1]: *** [Makefile:105: /home/stefano/repos/coconut/svsm-review/libvtpm/ms-tpm-20-ref/TPMCmd/Makefile] Error 1
make[1]: Leaving directory '/home/stefano/repos/coconut/svsm-review/libvtpm'
make: *** [Makefile:70: libvtpm/libvtpm.a] Error 2
```

Maybe adding a dependency will help:
```suggestion
libvtpm/bindings.rs: libvtpm/libvtpm.a
```

But I'm not sure it is fully right or just a workaround.

Note: I found another issue running `make` with multiple jobs, but not related to this PR, so I opened #247

cclaudio replied:
`make -j 10` is working for me now. I'd say that the latest changes  I added fixed this.

zvonkok commented:
/cc @zvonkok

cclaudio commented:
Added support for cargo workspace and rebased.
Updated the PR description.

The github rust.yml test is failing because `cargo test --workspace` is not honoring `--no-default-features`, see the last patch in the PR. I did not find a workaround for this cargo issue yet.

stefano-garzarella commented on libmstpm/Makefile:
```diff
@@ -0,0 +1,141 @@
+	$(MAKE) -C $(MSTPM_DIR) $(LIBPLATFORM_A)
+
+MSTPM_CFLAGS := -static -nostdinc -fno-stack-protector -fPIE -mno-sse
+MSTPM_CFLAGS += -DSIMULATION=NO -DFILE_BACKED_NV=NO
```
Playing with ms-tpm, I realized that `DEBUG=YES` is the default, so should we add `-DDEBUG=NO` here?

cclaudio replied:
Yes, I can add that. Thanks!

cclaudio replied:
Done.

cclaudio commented on libvtpm/libcrt/src/prng/rand.c:
```diff
@@ -0,0 +1,44 @@
+{
+  uint64_t r = 0;
+
+  if (rdrand64(&r)) {
```
Done.

cclaudio commented:
- Applied feedback for getentropy() in libmstpm/deps/libcrt/src/prng/rand.c
- Fixed `make test`

joergroedel commented on Documentation/INSTALL.md:
```diff
@@ -139,15 +139,16 @@ openSUSE you can do this by:
 $ make -j16 -C BaseTools/
-$ source ./edksetup.sh
-$ build -a X64 -b DEBUG -t GCC5 -D DEBUG_ON_SERIAL_PORT -D DEBUG_VERBOSE -p OvmfPkg/OvmfPkgX64.dsc
+$ source ./edksetup.sh --reconfig
```
Why is the --reconfig needed?

cclaudio replied:
Recently I decided to move my edk2 working tree to another folder. If we don't pass `--reconfig` the edk2 build will use the old configuration because the related env variables were not updated.

joergroedel commented on libmstpm/Makefile:
```diff
@@ -0,0 +1,141 @@
+BINDGEN_FLAGS = --use-core
+CLANG_FLAGS = -Wno-incompatible-library-redeclaration
+
+$(BINDGEN):
+	cargo install bindgen-cli@0.69
```
Please do not install build dependencies on Makefiles. The right way to do this is to just call `cargo bindgen` and let the user install it (e.g. via distro package manager). I think changing this also requires updates to the documentation about which build dependencies to install and the docker-files.

cclaudio replied:
I will do that, thanks.

stefano-garzarella replied:
If you will remove it, please ignore this comment ;-)

It's failing here:
```
$ make src/bindings.rs
cargo install bindgen-cli@0.69
error: invalid value 'bindgen-cli@0.69' for '[CRATE[@<VER>]]...': unexpected end of input while parsing minor version number

  tip: if you want to specify SemVer range, add an explicit qualifier, like '^0.69'

For more information, try '--help'.
make: *** [Makefile:119: /home/stefano/.cargo/bin/bindgen] Error 1
```

I think we should use `cargo install bindgen-cli@^0.69` or `cargo install bindgen-cli@~0.69`

cclaudio replied:
I will document that in the INSTALL.md, thanks.

joergroedel reviewed (changes requested):
In general this looks good to me, my testing also went well. Just found too minor odd things which need clarification/update.

deeglaze commented on Documentation/INSTALL.md:
````diff
@@ -189,6 +193,15 @@ Building the SVSM itself requires:
+```
+$ sudo zypper in system-user-mail make gcc curl patterns-devel-base-devel_basis \
+      glibc-devel-static git libclang13 autoconf autoconf-archive pkg-config \
+      automake libopenssl-devel
````
I don't think we should be leaving the C library dependencies to open to the toolchain container. These need to be part of SVSM's own dependency management.

osteffenrh replied:
You mean pull the sources for all dependencies into the repo?

Not sure what the overall plan is, if distros will do their own builds?
At least for Fedora packaging, we'd replace all of those deps with the Fedora provided ones anyway. Same for Rust crates, as far as possible.
And also build on Fedora. I see the container more as a kown-good dev env + CI thing?

I admit reproducible builds would be awesome. Not sure how this fits into the picture right now.

deeglaze commented on kernel/src/protocols/mod.rs:
```diff
@@ -6,11 +6,14 @@
 
 pub mod core;
 pub mod errors;
+#[cfg(feature = "mstpm")]
```
Is it possible to allow the build to have the tpm linked in but have a qemufwcfg file specify whether to make it discoverable? We want to have a single binary to release but still allow users to decide whether or not to attach a vTPM to their instance.

00xc commented on kernel/src/vtpm/mstpm/mod.rs:
```diff
@@ -0,0 +1,192 @@
+            return Err(SvsmReqError::invalid_parameter());
+        }
+
+        let mut request_ffi = Vec::<u8>::with_capacity(*length);
+        request_ffi.extend_from_slice(&buffer[..*length]);
```
```suggestion
        let mut request_ffi = buffer[..*length].to_vec();
```

cclaudio replied:
True :-). I will do that, thanks.

00xc commented on kernel/src/svsm.rs:
```diff
@@ -441,6 +443,11 @@ pub extern "C" fn svsm_main() {
     }
 
+    #[cfg(feature = "mstpm")]
+    if let Err(e) = vtpm_init() {
+        panic!("vTPM failed to initialize - {:?}", e);
+    }
```
This will also display the error and is shorter:

```suggestion
    vtpm_init().expect("vTPM failed to initialize");
```

cclaudio replied:
I will do that, thanks.

00xc commented on kernel/src/protocols/vtpm.rs:
```diff
@@ -0,0 +1,220 @@
+        let mut buffer: Vec<u8> = Vec::with_capacity(SEND_COMMAND_RESP_OUTBUF_SIZE);
+        buffer.extend_from_slice(tpm_cmd);
+
+        // The buffer slice must be large enough to hold the TPM command response
+        unsafe { buffer.set_len(buffer.capacity()) };
```
This creates a reference to uninitialized memory and is undefined behavior, even if memory is not accessed.

```suggestion
        // The buffer slice must be large enough to hold the TPM command response
        buffer.resize(SEND_COMMAND_RESP_OUTBUF_SIZE, 0);
```

00xc commented on kernel/src/protocols/vtpm.rs:
```diff
@@ -0,0 +1,220 @@
+        let vtpm = vtpm_get_locked();
+        vtpm.send_tpm_command(buffer.as_mut_slice(), &mut length, self.locality)?;
+
+        unsafe {
+            if length > buffer.capacity() {
+                return Err(SvsmReqError::invalid_request());
+            }
+            buffer.set_len(length);
+        }
```
Again, we cannot set the length to capacity because that could include allocated but uninitialized memory (note that capacity might be greater than `SEND_COMMAND_RESP_OUTBUF_SIZE`, we are only guaranteed that's the minimum).

With my previous suggestion we filled up buffer until `SEND_COMMAND_RESP_OUTBUF_SIZE` with zeroes, so we can simply check that the new length is equal or smaller, and then truncate, avoiding unsafe completely.

```suggestion
        if length > buffer.len() {
            return Err(SvsmReqError::invalid_request());
        }
        buffer.truncate(length);
```

00xc commented on kernel/src/protocols/vtpm.rs:
```diff
@@ -0,0 +1,220 @@
+    Ok(())
+}
+
+fn tpm_send_command_request(iobuf: VirtAddr) -> Result<u32, SvsmReqError> {
+    let outbuf: Vec<u8> = {
+        let request = unsafe { &*iobuf.as_mut_ptr::<TpmSendCommandRequest>() };
+        request.send()?
+    };
+    let response = unsafe { &mut *iobuf.as_mut_ptr::<TpmSendCommandResponse>() };
+    let _ = response.set_outbuf(outbuf.as_slice());
+
+    Ok(outbuf.len() as u32)
+}
```
Please mark this function as unsafe - the caller must make sure that `iobuf` points to a sufficiently large region of memory, that it's properly aligned, that it is initialized to a valid representation of `TpmSendCommandResponse` and that it's not aliased.

cclaudio replied:
@00xc I did some improvements to this function.

```rust
/// Send a TpmSendCommandRequest to the vTPM
/// 
/// # Arguments
/// 
/// * `buffer`: Contains the TpmSendCommandRequest. It will also be
///             used to store the TpmSendCommandResponse as a byte slice
/// 
/// # Returns
/// 
/// * `u32`: Number of bytes written back to `buffer` as part of
///          the TpmSendCommandResponse
fn tpm_send_command_request(buffer: &mut [u8]) -> Result<u32, SvsmReqError> {
    let outbuf: Vec<u8> = {
        let request = TpmSendCommandRequest::try_from_as_ref(buffer)?;
        request.send()?
    };
    let response = TpmSendCommandResponse::try_from_as_mut_ref(buffer)?;
    let _ = response.set_outbuf(outbuf.as_slice());

    Ok(outbuf.len() as u32)
}
```
The caller now passes a mutable byte slice and the casting safety is checked inside the `try_from_as*` functions, which are similar to the one we have in `greq/pld_report.rs`

roy-hopkins commented on Documentation/INSTALL.md:
````diff
@@ -189,6 +193,15 @@ Building the SVSM itself requires:
+      automake libopenssl-devel
+```
+
 Then checkout the SVSM repository and build the SVSM binary:
 
 ```
````
I may have missed it elsewhere in the documentation but do we need to install git submodules now when cloning the SVSM repository?

```
$ git clone https://github.com/coconut-svsm/svsm
$ git submodule update --init
```

stefano-garzarella replied:
@cclaudio Am I remembering wrong or in the previous version of this PR we did it in the Makefile?
BTW, could it make sense to do it in the Makefile?

00xc commented on kernel/src/vtpm/mod.rs:
```diff
@@ -85,3 +110,8 @@ pub fn vtpm_init() -> Result<(), SvsmReqError> {
     Ok(())
 }
+
+pub fn vtpm_get_locked() -> VtpmRef {
+    let vtpm = addr_of_mut!(*VTPM.lock());
+    VtpmRef::new(vtpm)
+}
```
This is dropping the lock as soon as the function returns, which will lead to undefined behavior when using the `VtpmRef`. You need `VtpmRef` to hold a `SpinLockGuard<'a, Vtpm>` (or directly return that type from `vtpm_get_locked()`, which already implements `deref()` and `deref_mut()`).

cclaudio replied:
Returning `VTPM.lock()` directly from `vtpm_get_locked()` looks cleaner. 
`VtpmRef` is not needed anymore, deleted (along with its Deref implementations).
Thank you.

roy-hopkins commented on kernel/src/vtpm/mstpm/mod.rs:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    fn init(&mut self) -> Result<(), SvsmReqError> {
+        // Initialize the MS TPM following the same steps done in the Simulator:
+        //
+        // 1. Manufacture it for the first time
+        // 2. Make sure it does not fail if it is re-manufactured
+        // 3. Teardown to indicate it needs to be manufactured
+        // 4. Manufacture it for the first time
+        // 5. Power it on indicating it requires startup. By default, OVMF will start
+        //    and selftest it.
```
What is the purpose of this initialisation process? Maybe this should be moved to a `test-in-svsm` test and the startup process here simplified?

cclaudio replied:
IIUC, the MS TPM Simulator goes through [these steps](https://github.com/microsoft/ms-tpm-20-ref/blob/main/TPMCmd/Simulator/src/TPMCmds.c#L267) to be compliant with the TPM specs. We are just following the same steps.

stefano-garzarella commented on Makefile:
```diff
@@ -131,5 +131,4 @@ clean:
 	rm -f ${STAGE1_OBJS} utils/gen_meta utils/print-meta
 	rm -rf bin
 
-.PHONY: test clean clippy bin/stage2.bin bin/svsm-kernel.elf bin/test-kernel.elf
```
Should we add `distclean` in `.PHONY` in the same commit where we define `distclean` target?
(Maybe something went wrong with the rebase)

stefano-garzarella commented on libmstpm/build.rs:
```diff
@@ -0,0 +1,11 @@
+use std::process::Command;
+
+fn main() {
+    Command::new("make").status().unwrap();
```
If `make` fails here, is there any way to propagate the output?

Even using `make V=2` in the root, I can't see the errors produced by this call (they might be useful, for example in my case it failed because it can't install `bindgen-cli`, but I discovered it only doing `make -C libmstpm ` )

cclaudio replied:
Fixed, I patched `libmstpm/build.rs` to panic and print the stderr if it fails.

stefano-garzarella replied:
@cclaudio great! Thanks for fixing this!

osteffenrh commented:
I tried building with `scripts/container/build.sh` and it fails:
```
...
   Compiling bootlib v0.1.0 (/home/osteffen/work/svsm/coconut/svsm/bootlib)
   Compiling igvm_defs v0.1.3
error[E0583]: file not found for module `bindings`
  --> libmstpm/src/lib.rs:13:1
   |
13 | pub mod bindings;
   | ^^^^^^^^^^^^^^^^^
   |
   = help: to create the module `bindings`, create file "libmstpm/src/bindings.rs" or "libmstpm/src/bindings/mod.rs"
   = note: if there is a `mod bindings` elsewhere in the crate already, import it with `use crate::...` instead

For more information about this error, try `rustc --explain E0583`.
error: could not compile `libmstpm` (lib) due to 1 previous error
make: *** [Makefile:91: bin/stage2.bin] Error 101
```
This is probably the mstpm failing, due to missing dependencies. Error messages hidden by build.rs?

Would be nice if the containerized build would work.

roy-hopkins commented on libmstpm/Makefile:
```diff
@@ -0,0 +1,141 @@
+			no-asm \
+			--with-rand-seed=getrandom \
+			-I$(LIBCRT_DIR)/include \
+			-Wl,rpath=$(LIBCRT_DIR) -lcrt -g -O0)
```
Debug information is turned on and optimisation level set to 0 here. Should optimisation be enabled?

cclaudio replied:
I changed this to follow the RELEASE variable found in the Makefile. If it is set all libmstpm are built with `-O3` otherwise `-g -O0`

osteffenrh commented:
> I tried building with `scripts/container/build.sh` and it fails:
> 
> ```
> ...
>    Compiling bootlib v0.1.0 (/home/osteffen/work/svsm/coconut/svsm/bootlib)
>    Compiling igvm_defs v0.1.3
> error[E0583]: file not found for module `bindings`
>   --> libmstpm/src/lib.rs:13:1
>    |
> 13 | pub mod bindings;
>    | ^^^^^^^^^^^^^^^^^
>    |
>    = help: to create the module `bindings`, create file "libmstpm/src/bindings.rs" or "libmstpm/src/bindings/mod.rs"
>    = note: if there is a `mod bindings` elsewhere in the crate already, import it with `use crate::...` instead
> 
> For more information about this error, try `rustc --explain E0583`.
> error: could not compile `libmstpm` (lib) due to 1 previous error
> make: *** [Makefile:91: bin/stage2.bin] Error 101
> ```
> 
> This is probably the mstpm failing, due to missing dependencies. Error messages hidden by build.rs?
> 
> Would be nice if the containerized build would work.

```diff
diff --git a/scripts/container/opensuse-rust.docker b/scripts/container/opensuse-rust.docker
index 621d107..27cd8d2 100644
--- a/scripts/container/opensuse-rust.docker
+++ b/scripts/container/opensuse-rust.docker
@@ -21,7 +21,7 @@ SHELL ["/bin/bash", "-c"]
 RUN zypper ref && \
     zypper install -y system-user-mail make gcc curl \
         patterns-devel-base-devel_basis glibc-devel-static git libclang13 \
-        autoconf autoconf-archive pkg-config automake libopenssl-devel && \
+        autoconf autoconf-archive pkg-config automake libopenssl-devel perl && \
     useradd -u $USER_ID -m $USER_NAME

 USER $USER_NAME
 ``` 
That seems to fix it.

cclaudio commented:
All comments addressed. Just waiting for #305 

- `make V=2` now should show all libmstpm build messages.
- If libmstpm fails to build, it will panic and print the stderr and stdout in context
- The libmstpm build is now honoring the RELEASE variable: if set, the libmstpm is built with `-O3`, otherwise `-g -O0`.

This PR is working for me, but if we want the libmstpm to be built with any optimization (`RELEASE=1 make`), we will need to fix the errors below by our first release:
1) `RELEASE=1 make` is failing. For some reason, the linker is complaining about undefined vtpm symbols that are actually defined.
2) Without passing `RELEASE=1`, if we force any optimization with `-O` in CFLAGS in the `openssl_svsm.conf`, the libmstpm builds fine but the guest fails to boot
````
[SVSM] ERROR: Panic: CPU[0] panicked at kernel/src/cpu/idt/svsm.rs:165:5:
Unhandled exception 6 RIP 0xffffff800001e2f0 error code: 0x0000000000000000
[SVSM] ---BACKTRACE---:
[SVSM]   [0xffffff800000015e]
[SVSM]   Invalid frame
[SVSM] ---END---
````

osteffenrh commented:
@cclaudio, I still have some smaller things around the containerized build. The patches below make it work for me.
(I tried attaching this as a .patch file but GH refuses it, although it claims this filetype is allowed... :thinking: )

```patch
From 40adf6651df48bdd58253436154d1e70eb1188c5 Mon Sep 17 00:00:00 2001
From: Oliver Steffen <osteffen@redhat.com>
Date: Mon, 25 Mar 2024 12:09:36 +0100
Subject: [PATCH 1/3] scripts/container: Set CARGO env in Dockerfile

Set CARGO_HOME and add it to PATH in the Dockerfile.
This way we don't have to source the setup script every time.

Signed-off-by: Oliver Steffen <osteffen@redhat.com>
---
 scripts/container/build.sh             | 4 ++--
 scripts/container/opensuse-rust.docker | 8 +++++++-
 2 files changed, 9 insertions(+), 3 deletions(-)

diff --git a/scripts/container/build.sh b/scripts/container/build.sh
index c2f2132..a907721 100755
--- a/scripts/container/build.sh
+++ b/scripts/container/build.sh
@@ -96,7 +96,7 @@ BuildSvsmReuse()
 
     "${DOCKER_CMD}" exec \
         -it ${CONTAINER_NAME} \
-        /bin/bash -c "source $HOME/.cargo/env && make clean && make"
+        /bin/bash -c "make clean && make"
 }
 
 ####
@@ -111,7 +111,7 @@ BuildSvsmDelete()
           --mount type=bind,source="${WORKDIR}",target="${WORKDIR}"${MOUNT_OPTS} \
           $EXTRA_DOCKER_OPTS \
           ${IMAGE_NAME} \
-          /bin/bash -c "source $HOME/.cargo/env && make clean && make"
+          /bin/bash -c "make clean && make"
 }
 
 ####
diff --git a/scripts/container/opensuse-rust.docker b/scripts/container/opensuse-rust.docker
index 27cd8d2..b1abe85 100644
--- a/scripts/container/opensuse-rust.docker
+++ b/scripts/container/opensuse-rust.docker
@@ -16,15 +16,21 @@ LABEL Description="OpenSUSE environment for coconut-svsm build"
 ARG USER_ID 1000
 ARG USER_NAME user
 
+ENV CARGO_HOME=/opt/cargo
+ENV RUSTUP_HOME=/opt/rustup
+
 SHELL ["/bin/bash", "-c"]
 
 RUN zypper ref && \
     zypper install -y system-user-mail make gcc curl \
         patterns-devel-base-devel_basis glibc-devel-static git libclang13 \
         autoconf autoconf-archive pkg-config automake libopenssl-devel perl && \
-    useradd -u $USER_ID -m $USER_NAME
+    useradd -u $USER_ID -m $USER_NAME && \
+       mkdir -p "${CARGO_HOME}" "${RUSTUP_HOME}" && \
+       chown "${USER_NAME}" "${CARGO_HOME}" "${RUSTUP_HOME}"
 
 USER $USER_NAME
+ENV PATH="${PATH}:${CARGO_HOME}/bin"
 
 RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs > /tmp/rustup-init.sh && \
     sh /tmp/rustup-init.sh -y \
-- 
2.44.0


From fbec5e8eeb9ecdf43e2a5232c3076baa9019fc9f Mon Sep 17 00:00:00 2001
From: Oliver Steffen <osteffen@redhat.com>
Date: Mon, 25 Mar 2024 12:11:38 +0100
Subject: [PATCH 2/3] scripts/container: Add bindgen-cli

Install bindgen-cli in container image.

Signed-off-by: Oliver Steffen <osteffen@redhat.com>
---
 scripts/container/opensuse-rust.docker | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/scripts/container/opensuse-rust.docker b/scripts/container/opensuse-rust.docker
index b1abe85..bf44a68 100644
--- a/scripts/container/opensuse-rust.docker
+++ b/scripts/container/opensuse-rust.docker
@@ -35,4 +35,5 @@ ENV PATH="${PATH}:${CARGO_HOME}/bin"
 RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs > /tmp/rustup-init.sh && \
     sh /tmp/rustup-init.sh -y \
         --default-toolchain stable-x86_64-unknown-linux-gnu \
-        --target x86_64-unknown-none
+        --target x86_64-unknown-none && \
+    cargo install bindgen-cli
-- 
2.44.0


From 6734890ac195c533b8022e98155dcc29607d8537 Mon Sep 17 00:00:00 2001
From: Oliver Steffen <osteffen@redhat.com>
Date: Mon, 25 Mar 2024 12:13:53 +0100
Subject: [PATCH 3/3] scripts/container: Fix usage help

Mention submodule initialization in the comment exlaining
how to use the container build scripts.

Signed-off-by: Oliver Steffen <osteffen@redhat.com>
---
 scripts/container/opensuse-rust.docker | 1 +
 1 file changed, 1 insertion(+)

diff --git a/scripts/container/opensuse-rust.docker b/scripts/container/opensuse-rust.docker
index bf44a68..483105c 100644
--- a/scripts/container/opensuse-rust.docker
+++ b/scripts/container/opensuse-rust.docker
@@ -8,6 +8,7 @@
 #
 # git clone https://github.com/coconut-svsm/svsm.git
 # cd svsm
+# git submodule update --init --depth=1
 # ./scripts/docker/build.sh 
 
 FROM opensuse/tumbleweed:latest
-- 
2.44.0


```

cclaudio commented:
> @cclaudio, I still have some smaller things around the containerized build. The patches below make it work for me. (I tried attaching this as a .patch file but GH refuses it, although it claims this filetype is allowed... 🤔 )
> 

@osteffenrh `build.sh` is working fine for me. What issue are you having?
You may need to do the clean-up below before running the `build.sh` 

```
$ cd scripts/container
$ docker stop coconut-build  # if a coconut-build container exists
$ docker rm  coconut-build  # if a coconut-build container exists
$ docker image rm opensuse-rust
$ ./build.sh
```

osteffenrh commented:
> > @cclaudio, I still have some smaller things around the containerized build. The patches below make it work for me. (I tried attaching this as a .patch file but GH refuses it, although it claims this filetype is allowed... 🤔 )
> 
> @osteffenrh `build.sh` is working fine for me. What issue are you having? You may need to do the clean-up below before running the `build.sh`
> 
> ```
> $ cd scripts/container
> $ docker stop coconut-build  # if a coconut-build container exists
> $ docker rm  coconut-build  # if a coconut-build container exists
> $ docker image rm opensuse-rust
> $ ./build.sh
> ```

bindgen is missing in the container image. That is the main thing the patches above change.

I just did a fresh clone of your branch and it fails like this (no images/containers pres:

```
$ podman ps
CONTAINER ID  IMAGE       COMMAND     CREATED     STATUS      PORTS       NAMES
╭─osteffen@virtlab1020 ~
╰─$ podman ps -a
CONTAINER ID  IMAGE       COMMAND     CREATED     STATUS      PORTS       NAMES
$ podman images                                                                          125 ↵
REPOSITORY                                 TAG         IMAGE ID      CREATED       SIZE
registry.opensuse.org/opensuse/tumbleweed  latest      eee52e78a63a  3 days ago    142 MB
registry.fedoraproject.org/fedora          39          368a084ba17d  3 months ago  183 MB
$ git clone https://github.com/cclaudio/svsm.git --branch coconut-svsm-vtpm --depth=1 --single-branch
Cloning into 'svsm'...
remote: Enumerating objects: 297, done.
remote: Counting objects: 100% (297/297), done.
remote: Compressing objects: 100% (278/278), done.
remote: Total 297 (delta 10), reused 184 (delta 8), pack-reused 0
Receiving objects: 100% (297/297), 307.19 KiB | 4.65 MiB/s, done.
Resolving deltas: 100% (10/10), done.
$ git submodule update --init --depth=1
Submodule 'libmstpm/deps/ms-tpm-20-ref' (https://github.com/microsoft/ms-tpm-20-ref.git) registered for path 'libmstpm/deps/ms-tpm-20-ref'
Submodule 'libmstpm/deps/openssl' (https://github.com/openssl/openssl.git) registered for path 'libmstpm/deps/openssl'
Cloning into '/tmp/vtpm/svsm/libmstpm/deps/ms-tpm-20-ref'...
Cloning into '/tmp/vtpm/svsm/libmstpm/deps/openssl'...
Submodule path 'libmstpm/deps/ms-tpm-20-ref': checked out 'e9fc7b89d865536c46deb63f9c7d0121a3ded49c'
remote: Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
remote: Enumerating objects: 19975, done.
remote: Counting objects: 100% (19975/19975), done.
remote: Compressing objects: 100% (14747/14747), done.
remote: Total 17797 (delta 2481), reused 14878 (delta 534), pack-reused 0
Receiving objects: 100% (17797/17797), 8.47 MiB | 10.44 MiB/s, done.
Resolving deltas: 100% (2481/2481), completed with 1846 local objects.
From https://github.com/openssl/openssl
 * branch              29708a562a1887a91de0fa6ca668c71871accde9 -> FETCH_HEAD
Submodule path 'libmstpm/deps/openssl': checked out '29708a562a1887a91de0fa6ca668c71871accde9'
$ cd scripts/container
$ build.sh
[...]
  configure.ac:42: installing './missing'
  Makefile.am: installing './depcomp'
  /bin/sh: line 1: bindgen: command not found
  make[1]: *** [Makefile:132: src/bindings.rs] Error 127
  thread 'main' panicked at libmstpm/build.rs:18:9:
  explicit panic
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
make: *** [Makefile:91: bin/stage2.bin] Error 101
```

Full log attached. ([buildlog.txt](https://github.com/coconut-svsm/svsm/files/14747356/buildlog.txt))

It might be that you are not seeing the issue because of left-over build artifacts in libmstpm ? Try `git clean -xdf && git submodule foreach git clean -xdf`?

cclaudio commented:
True, I am fixing that. Thanks @osteffenrh

cclaudio commented:
@deeglaze, containerized builds is still under discussion in #287, I am not sure what we should do here.  Should we address it in a separate PR? The problem does not seem to be specific to this one.

All other comments addressed

cclaudio commented:
rebased and fixed conflict with Makefile

joergroedel commented:
This is merged now. Thanks @cclaudio  and all the reviewers!
